syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,987 @@
+
+    return with_session_wrapper
+
+
+def user_key(user):
+    """Create a compound user-key using the fxa_uid and fxa_kid
+
+    Joined here rather than used as a compound primary key to allow re-use
+    of existing queries.
+
+    """
+    return ":".join([user["fxa_uid"], user["fxa_kid"]])
+
+
+def getq(query):
+    kwargs = {}
+    if "%(bso)s" in query:
+        kwargs["bso"] = "bso"
+    if "%(bui)s" in query:
+        raise RuntimeError("XXX: batches")
+    if "%(ids)s" in query:
+        kwargs["ids"] = "@ids"
+    if kwargs:
+        query = query % kwargs
+    return query.replace(":", "@")
+
+
+class SpannerStorage(SyncStorage):
+    """Storage plugin implemented using a Spanner database.
+
+    This class implements the storage plugin API using Google Spanner.  You
+    must specify the Spanner database URI string to connect to, as follows:
+
+        spanner://INSTANCE_ID:DATABASE_ID
+
+    """
+
+    def __init__(self, sqluri, standard_collections=False, **dbkwds):
+        self.sqluri = sqluri
+        instance_id, database_id = sqluri[len("spanner://"):].split(":")
+        self.client = spanner.Client()
+        self._instance = self.client.instance(instance_id)
+        self._pool = pool = spanner.BurstyPool()
+        self._database = self._instance.database(database_id, pool=pool)
+
+        self.standard_collections = standard_collections
+        if self.standard_collections and dbkwds.get("create_tables", False):
+            raise Exception("Creating tables in Spanner must be done manually")
+
+        # A local in-memory cache for the name => collectionid mapping.
+        self._collections_by_name = {}
+        self._collections_by_id = {}
+        if self.standard_collections:
+            for id, name in STANDARD_COLLECTIONS.iteritems():
+                self._collections_by_name[name] = id
+                self._collections_by_id[id] = name
+
+        # A thread-local to track active sessions.
+        self._tldata = threading.local()
+
+    def _lock_write_precondition(self, session, user, collection):
+        """'Lock' the precondition on the collection
+
+        Get the timestamp of the collection in use, utilizing it as
+        precondition for
+        transaction invalidation.
+
+        """
+        result = session.transaction.execute_sql(
+            COLLECTION_CURRENT_TIMESTAMP,
+            params={"userid": user_key(user),
+                    "collectionid": collection},
+            param_types={"userid": param_types.STRING,
+                         "collectionid": param_types.INT64}
+        ).one_or_none()
+        if result:
+            current_ts, last_modified = result
+            session.collection_ts = dt2ts(last_modified)
+            session.timestamp_precise = current_ts
+            session.timestamp = dt2ts(current_ts)
+            # Forbid the write if it would not properly incr the timestamp.
+            if session.collection_ts >= dt2ts(current_ts):
```

Comment:
nit: you could just use `session.timestamp` here rather than calling `dt2ts` twice.